demos/http3/ossl-nghttp3.c

```diff
@@ -0,0 +1,603 @@
+        goto err;
+
+    if (settings == NULL) {
+        nghttp3_settings_default(&dsettings);
+        settings = &dsettings;
+    }
+
+    if (callbacks != NULL)
+        intl_callbacks = *callbacks;
+
+    conn->recv_data_cb          = intl_callbacks.recv_data;
+    conn->stream_close_cb       = intl_callbacks.stream_close;
+    conn->stop_sending_cb       = intl_callbacks.stop_sending;
+    conn->reset_stream_cb       = intl_callbacks.reset_stream;
+    conn->deferred_consume_cb   = intl_callbacks.deferred_consume;
+
+    intl_callbacks.recv_data        = h3_conn_recv_data;
+    intl_callbacks.stream_close     = h3_conn_stream_close;
+    intl_callbacks.stop_sending     = h3_conn_stop_sending;
+    intl_callbacks.reset_stream     = h3_conn_reset_stream;
+    intl_callbacks.deferred_consume = h3_conn_deferred_consume;
+
+    ec = nghttp3_conn_client_new(&conn->h3conn, &intl_callbacks, settings,
+                                 NULL, conn);
+    if (ec < 0) {
+        ERR_raise_data(ERR_LIB_USER, ERR_R_INTERNAL_ERROR,
+                       "cannot create nghttp3 connection: %s (%d)",
+                       nghttp3_strerror(ec), ec);
+        goto err;
+    }
+
+    ec = nghttp3_conn_bind_control_stream(conn->h3conn, s_ctl_send->id);
+    if (ec < 0) {
+        ERR_raise_data(ERR_LIB_USER, ERR_R_INTERNAL_ERROR,
+                       "cannot bind nghttp3 control stream: %s (%d)",
+                       nghttp3_strerror(ec), ec);
+        goto err;
+    }
+
+    ec = nghttp3_conn_bind_qpack_streams(conn->h3conn,
+                                         s_qpenc_send->id,
+                                         s_qpdec_send->id);
+    if (ec < 0) {
+        ERR_raise_data(ERR_LIB_USER, ERR_R_INTERNAL_ERROR,
+                       "cannot bind nghttp3 QPACK streams: %s (%d)",
+                       nghttp3_strerror(ec), ec);
+        goto err;
+    }
+
+    return conn;
+
+err:
+    nghttp3_conn_del(conn->h3conn);
+    h3_stream_free(s_ctl_send);
+    h3_stream_free(s_qpenc_send);
+    h3_stream_free(s_qpdec_send);
+    lh_H3_STREAM_free(conn->streams);
+    OPENSSL_free(conn);
+    return NULL;
+}
+
+H3_CONN *H3_CONN_new_for_addr(SSL_CTX *ctx, const char *addr,
+                              const nghttp3_callbacks *callbacks,
+                              const nghttp3_settings *settings,
+                              void *user_data)
+{
+    BIO *qconn_bio = NULL;
+    SSL *qconn = NULL;
+    H3_CONN *conn = NULL;
+    const char *bare_hostname;
+
+    /* QUIC connection setup */
+    if ((qconn_bio = BIO_new_ssl_connect(ctx)) == NULL)
+        goto err;
+
+    if (BIO_set_conn_hostname(qconn_bio, addr) == 0)
+        goto err;
+
+    bare_hostname = BIO_get_conn_hostname(qconn_bio);
+    if (bare_hostname == NULL)
+        goto err;
+
+    if (BIO_get_ssl(qconn_bio, &qconn) == 0)
+        goto err;
+
+    if (SSL_set1_host(qconn, bare_hostname) <= 0)
```

Comment:
Shouldn't we also call `SSL_set_tlsext_host_name()` here?

Reply:
We do this in ddd-03 (e.g. direct FD) where it's necessary but it's not necessary when using the ssl_connect BIO, right?

TLS 1.3 requires SNI so it wouldn't even work without it, no...?